### PR TITLE
Escorts System

### PR DESCRIPTION
**Pull request by thvk-net**
-----------------------
**Escorts System:**
Well something like that: https://github.com/endless-sky/endless-sky/pull/6536

## Feature Details
This (draft) PR adds the possibility to add missions that work as escorts for the player. The will pay them on daily base and if he has not enough money the escorts will leave him.

One new tags for missions: 
- `"escort salary" [cost]` marks the mission as an escort mission and defines the daily cost.

All info on how to create a new escort can be found in escorts.txt in data/humans/

Important Info: This PR only adds the possibility of creating escorts it is not adding them.

### CODE REVIEW

**Comment by thvk-net** [comment]
Ignore the things to the coalition ok? They just slipped in somehow.

**Comment by petervdmeer** [comment]
Thanks for contributing, and sorry that we didn't get to your PR sooner.

Looks like this PR was created on top of a non-master branch, causing changes from another PR to show up within this PR.
A PR should only contain commits relevant for the change in the PR, so this needs to be fixed before we can review this PR.

Also please consider to first file an issue to discuss the plans for this PR; this PR seems to introduce 2 new keywords, and an issue is a good place to discuss if this is the direction to take or if we should do this differently.

File: source/Account.cpp [hunk 1]
```diff
@@ -136,6 +136,9 @@ string Account::Step(int64_t assets, int64_t salaries, int64_t maintenance)
 	maintenanceDue += maintenance;
 	bool missedPayment = false;
 
+	credits -= escortCosts;
+	out << "You paid " << escortCosts << " credits for escorts. ";
```
Comment (Hurleveur) [inline source/Account.cpp:140]: Won't this leave a message even if escortCosts is equal to 0?

File: source/Mission.cpp [hunk 2]
```diff
@@ -184,12 +185,16 @@ void Mission::Load(const DataNode &node)
 			// This was an "illegal" or "stealth" entry. It has already been
 			// parsed, so nothing more needs to be done here.
 		}
+		else if(child.Token(0) == "cost" && child.Size() >= 2)
```
Comment (Hurleveur) [inline source/Mission.cpp:188]: I would suggest renaming `cost` to `player escort daily cost` or `escort daily cost` or smth of the like.
The variable could simply be `dailyCost`

If we do it this way we can just have it be one clear keyword for both things; I don't think we intend to add the possibility of a mission solely with a daily cost here.

That said you can keep the IsPlayerEscort() method, and just have it check if we have a `dailyCost` != 0.

File: source/Mission.h [hunk 3]
```diff
@@ -81,6 +81,9 @@ class Mission {
 	// Check if this mission is a "minor" mission. Minor missions will only be
 	// offered if no other missions (minor or otherwise) are being offered.
 	bool IsMinor() const;
+	// Check if this mission is an player escort. If yes it will be handled
```
Comment (Hurleveur) [inline source/Mission.h:84]:
```suggestion
	// Check if this mission is a player escort. If yes it will be handled
```

File: source/Mission.cpp [hunk 4]
```diff
@@ -605,6 +624,12 @@ int Mission::Passengers() const
 
 
 
+int Mission::Cost() const
+{
+	return cost;
+}
+
+
```
Comment (Hurleveur) [inline source/Mission.cpp:632]:
```suggestion


```
You forgot one empty line here. (we don't consider the comment as a line)

File: source/Mission.h [hunk 5]
```diff
@@ -212,6 +217,8 @@ class Mission {
 	// Parameters for generating random passenger amounts:
 	int passengerLimit = 0;
 	double passengerProb = 0.;
+	// Escort payment
+	int cost = 0;
```
Comment (Hurleveur) [inline source/Mission.h:221]: The name of the variable could be clear enough not to require a comment; for instance `escortSalary` instead of cost

File: source/MissionPanel.cpp [hunk 6]
```diff
@@ -148,6 +152,7 @@ MissionPanel::MissionPanel(PlayerInfo &player)
 		DoScroll(accepted, acceptedIt, acceptedScroll, true);
 	}
 
+
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:155]:
```suggestion
```
This whitespace change is not needed

File: source/MissionPanel.cpp [hunk 7]
```diff
@@ -219,6 +227,7 @@ void MissionPanel::Draw()
 		system = next;
 	}
 
+	// Highlight systems with active or available miisions.
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:230]:
```suggestion
	// Highlight systems with active or available missions.
```
Small typo

File: source/MissionPanel.cpp [hunk 8]
```diff
@@ -355,13 +386,27 @@ bool MissionPanel::Click(int x, int y, int clicks)
...
+			index -= available.size() + 1;
+			availableEscortsIt = availableEscorts.begin();
+			while(index--)
+				++availableEscortsIt;
+			acceptedIt = accepted.end();
+			availableIt = available.end();
+			dragSide = -1;
+			DoScroll(availableEscorts, availableEscortsIt, availableEscortsScroll, false);
+			return true;
+		}
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:408]: I'm a bit confused by this, why do these missions need to be handled differently from the rest; why are they not in the other available missions?
The mission has a destination by default btw, if that's what you're worried about for centering around said destination (its the source, but it could be something else and it should probably be shown here like the rest)

File: source/MissionPanel.cpp [hunk 9]
```diff
@@ -355,13 +386,27 @@ bool MissionPanel::Click(int x, int y, int clicks)
...
+
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:409]:
```suggestion
```
Looks weird if we add a line here right before closing the braces

**Review by Hurleveur — CHANGES_REQUESTED** [review]
This opens some nice possibilities for roleplaying a merchant, I quite like it, however there a few small things that I would like to be addressed.

Instead of creating a file `escort.txt` you can just include the syntax to make one such mission within the description of the PR; that way the changes are kept to code only.

On the code itself I don't like the fact that availableEscorts is treated like an exception, I don't really see why it can't be a part of the other available missions.

**Comment by Hurleveur** [comment]
> Looks like this PR was created on top of a non-master branch, causing changes from another PR to show up within this PR.

@petervdmeer I just looked at the PR in great detail there is not any change that isn't related to it.
It looks like its just a few commits that got lost, which I don't think matters that much; this will all be made into one commit when we merge it into master anyway, and most importantly not everyone is a git expert, relevant commits start after the `Removed things that don't belong here` commit.

That said I fully agree that it should not be a habit to do it this way, but maybe we could let it slide here? There could be graver consequences I am not aware of though, but the other changes were only content wise so its easy to see none remain.
If you really want thvk-net to do it, they could be helped on how to? (I'm no git command line expert myself)

File: source/Account.cpp (hunk 1 above)
Comment (thvk-net) [reply source/Account.cpp:140]: Yes, you  are right I just didnt thought of an escort that has no cost

File: source/MissionPanel.cpp (hunk 8 above)
Comment (thvk-net) [reply source/MissionPanel.cpp:407]: Because of the extra heading I added on the job board

File: source/Mission.h (hunk 5 above)
Comment (thvk-net) [reply source/Mission.h:221]: Will do this

**Comment by thvk-net** [comment]
Well, thanks for your replies, I am sorry that I did not follow the steps github users are normally following, I am not using github often. A lot of the basic changes Hurleveur requested where applied. The other changes/request will be worked out now.

File: source/Mission.h (hunk 5 above)
Comment (thvk-net) [reply source/Mission.h]: Is done

**Comment by thvk-net** [comment]
I have applied the suggested changes with the newest commit.

**Comment by Hurleveur** [comment]
I think you misunderstood something for the syntax, what I meant is to change to the variables to `escortSalary` (which you did), however the mission syntax is different from the code syntax, it should be `"escort salary"`.
Btw in order to make it multiple words you just need to surround them with these `"`, `escort salary` is considered as two separate things, and `"escort salary"` is read as one.

File: source/MissionPanel.cpp (hunk 8 above)
Comment (Hurleveur) [reply source/MissionPanel.cpp:407]: Oooh right I just noticed that, good thinking.

From a player's perspective, it makes sense to show them separate for sure, but code-wise maybe we could still have them in the same vector; at the end of it, and when we display missions we do two passes, one skipping these and then another one only considering them?
I would like to have another opinion on how it should be done, I'm not sure.

Comment (Hurleveur) [reply source/MissionPanel.cpp:407]: It seems like available jobs are simply sorted alphabetically, maybe they could be sorted based on if its an escort job and then alphabetically, that way we would know that once we find one escort mission the rest will be escorts too.
That said I'm still not sure that is the best way to do it

File: data/human/escorts.txt [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+#		government "Republic" "Free Worlds" "Syndicate" "Neutral" "Independent"
+#	to complete
+#		never
+#	npc accompany save
+#		government "Merchant"
+#		personality escort heroic
+#		fleet
+#			names "civilian"
+#			variant
+#				"Interceptor"
```
Comment (Hurleveur) [inline data/human/escorts.txt:19]:
```suggestion
#mission "Escort"
#	name "Escort"
#	description "This Captain will Escort you for 200 credits per day."
#	repeat
#	job
#	"escort salary" 200
#	source
#		government "Republic" "Free Worlds" "Syndicate" "Neutral" "Independent"
#	to complete
#		never
#	npc accompany save
#		government "Merchant"
#		personality escort heroic
#		fleet
#			names "civilian"
#			variant
#				"Sparrow"
```
This should be moved into PR description but since its here, let me fix it first.

File: source/Mission.cpp [hunk 11]
```diff
@@ -184,6 +185,8 @@ void Mission::Load(const DataNode &node)
 			// This was an "illegal" or "stealth" entry. It has already been
 			// parsed, so nothing more needs to be done here.
 		}
+		else if(child.Token(0) == "escortSalary" && child.Size() >= 2)
```
Comment (Hurleveur) [inline source/Mission.cpp:188]:
```suggestion
		else if(child.Token(0) == "escort salary" && child.Size() >= 2)
```

File: source/Mission.cpp [hunk 12]
```diff
@@ -331,6 +334,8 @@ void Mission::Save(DataWriter &out, const string &tag) const
 			out.Write("priority");
 		if(isMinor)
 			out.Write("minor");
+		if(escortSalary)
+			out.Write("escortSalary", escortSalary);
```
Comment (Hurleveur) [inline source/Mission.cpp:338]:
```suggestion
			out.Write("escort salary", escortSalary);
```

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Just to be clear, this is what I mean; the syntax is different in the code and in the missions; in the code variables use UpperCase to separate them whereas in the missions we use "normal case"

File: data/human/escorts.txt (hunk 10 above)
Comment (Hurleveur) [reply data/human/escorts.txt:19]: Under variant you are supposed to specify the ship type, you can't just type Interceptor there though
There is a wiki (on the bar up top with pull request etc, third one from the right, you can easily find syntax and all in there)

File: source/MissionPanel.cpp (hunk 8 above)
Comment (thvk-net) [reply source/MissionPanel.cpp:407]: This would be one way yes.

I will have some more thoughts on that point.

**Comment by petervdmeer** [comment]
> That said I fully agree that it should not be a habit to do it this way, but maybe we could let it slide here?

Sorry, no. We will not let it slide.
I understand that Git can be very unforgiving for non-Git experts, but we already have 150 open reviews and if we start accepting PRs that contain non-related commits, then the load on the reviewers will become even larger. It also helps for future reference if PRs contain the actual change and nothing else.

The actual set of Git commands to fix an issue like this can be very complex, but there are also simpler ways to fix this, for example;
1. Create a new escort2 branch
2. Squash-merge this branch into the escort2 branch (causing all 47 commits to already become 1 single commit, use a nice commit message here).
3. Clone this escort2 branch locally and force-push it back as the escort branch (which will reduce this PR to the 1 single commit just created)

This is a very rude way of fixing this issue, there are far more elegant (more complex) ways to do this, but I believe it will work. Be aware during squash-merging that Arachi-Lover will likely be listed as Co-Author (since Arachi-Lover is listed in the earlier commits). You will want to remove this Co-Author, since it looks like Arachi-Lover had nothing to do with the actual functionality in this PR.

Also make some backups when doing this; force-push is a very strong command and if you force-push to the wrong branch then you could actually lose data on the branch you are pushing to.

File: data/human/escorts.txt [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+#		government "Republic" "Free Worlds" "Syndicate" "Neutral" "Independent"
+#	to complete
+#		never
+#	npc accompany save
+#		government "Merchant"
+#		personality escort heroic
+#		fleet
+#			names "civilian"
+#			variant
+#				"Sparrow"
```
Comment (Hurleveur) [inline data/human/escorts.txt:19]: Could you move all of this into the PR's description?
That's where people will look how to use the new feature. If you're worried it'll take too much space, use this:
```
<details><summary>Usage example.</summary>

Here, after the blank line, put the example.
</details>```

File: source/Mission.cpp [hunk 14]
```diff
@@ -32,6 +32,7 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include <cmath>
 #include <sstream>
 
+
```
Comment (Hurleveur) [inline source/Mission.cpp:35]:
```suggestion
```
Unnecessary whitespace change

File: source/Account.cpp [hunk 15]
```diff
@@ -127,7 +127,7 @@ void Account::PayExtra(int mortgage, int64_t amount)
 
 
 // Step forward one day, and return a string summarizing payments made.
-string Account::Step(int64_t assets, int64_t salaries, int64_t maintenance)
+string Account::Step(int64_t assets, int64_t salaries, int64_t maintenance, int64_t escortCosts)
```
Comment (Hurleveur) [inline source/Account.cpp:130]:
```suggestion
string Account::Step(int64_t assets, int64_t salaries, int64_t maintenance, int64_t escortSalaries)
```
For coherence lets always use escortSalaries (you could also just use escortCosts everywhere if you prefer that but it needs to be coherent, I personally prefer salary because it reflects it is a daily cost even if you're not actually paying for their crew salaries, you're paying for the ship)

File: source/Account.cpp [hunk 16]
```diff
@@ -136,6 +136,10 @@ string Account::Step(int64_t assets, int64_t salaries, int64_t maintenance)
 	maintenanceDue += maintenance;
 	bool missedPayment = false;
 
+	credits -= escortCosts;
+	if(escortCosts != 0)
+		out << "You paid " << escortCosts << " credits for escorts. ";
```
Comment (Hurleveur) [inline source/Account.cpp:141]:
```suggestion
	credits -= escortSalaries;
	if(escortSalaries != 0)
		out << "You paid " << escortSalaries << " credits for escorts. ";
```
If you agree with changing everything to escortSalaries then it needs to be changed here as well

File: source/Account.h [hunk 17]
```diff
@@ -38,7 +38,7 @@ class Account {
 	void PayExtra(int mortgage, int64_t amount);
 
 	// Step forward one day, and return a string summarizing payments made.
-	std::string Step(int64_t assets, int64_t salaries, int64_t maintenance);
+	std::string Step(int64_t assets, int64_t salaries, int64_t maintenance, int64_t escortCosts);
```
Comment (Hurleveur) [inline source/Account.h:41]:
```suggestion
	std::string Step(int64_t assets, int64_t salaries, int64_t maintenance, int64_t escortSalaries);
```

File: source/PlayerInfo.h [hunk 18]
```diff
@@ -181,9 +181,10 @@ class PlayerInfo {
 	// Get mission information.
 	const std::list<Mission> &Missions() const;
 	const std::list<Mission> &AvailableJobs() const;
+	const std::list<Mission> &AvailableEscorts() const;
 	const Mission *ActiveBoardingMission() const;
 	void UpdateMissionNPCs();
-	void AcceptJob(const Mission &mission, UI *ui);
+	void AcceptJob(const Mission &mission, UI *ui, bool acceptEscort = false);
```
Comment (Hurleveur) [inline source/PlayerInfo.h:187]:
```suggestion
	void AcceptJob(const Mission &mission, UI *ui, bool istEscortJob = false);
```

File: source/PlayerInfo.cpp [hunk 19]
```diff
@@ -1578,18 +1592,34 @@ void PlayerInfo::UpdateMissionNPCs()
...
-		if(&*it == &mission)
-		{
-			cargo.AddMissionCargo(&mission);
-			it->Do(Mission::OFFER, *this);
-			it->Do(Mission::ACCEPT, *this, ui);
-			auto spliceIt = it->IsUnique() ? missions.begin() : missions.end();
-			missions.splice(spliceIt, availableJobs, it);
-			break;
-		}
+	if(!acceptEscort)
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:1597]:
```suggestion
	if(!istEscortJob)
```

File: source/PlayerInfo.cpp [hunk 20]
```diff
@@ -1578,18 +1592,34 @@ void PlayerInfo::UpdateMissionNPCs()
 
 
 // Accept the given job.
-void PlayerInfo::AcceptJob(const Mission &mission, UI *ui)
+void PlayerInfo::AcceptJob(const Mission &mission, UI *ui, bool acceptEscort)
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:1595]:
```suggestion
void PlayerInfo::AcceptJob(const Mission &mission, UI *ui, bool istEscortJob)
```
I think this name is better because acceptEscort seems to say we're only accepting escort jobs or something like that, whereas its just that the job selected is an escort job.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Some more suggestions, I recommend applying them as one; adding them all to the batch from the files changed tab.

**Comment by Hurleveur** [comment]
Also for the mission if you want you can also do a small plugin for it so that people can easily test but I think that's a bit too complex when all this PR adds syntax wise is the ability to use `escort job` with a daily fee specified.

In general its best to keep the PR to content changes/ code changes only.

**Comment by thvk-net** [comment]
I will close this pr because i am not able to remove this commits, even after i tried hard. I will open a new pr with only one commit.
